# 1. Exploratory Data Analysis

### 1.1 Import packages for analysis

In [1]:
import pandas as pd
import numpy as np
from scipy import stats
import statsmodels.api as sm
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

sns.set_style("white")

pd.set_option("display.max_columns",100)
pd.set_option("display.max_rows",100)

### 1.2 Load dataset as "med"

In [2]:
med = pd.read_csv("KaggleV2-May-2016.csv")

### 1.3 Get first impression on information contained in dataset 
    shape, column types, missing values

In [3]:
med.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
PatientId         110527 non-null float64
AppointmentID     110527 non-null int64
Gender            110527 non-null object
ScheduledDay      110527 non-null object
AppointmentDay    110527 non-null object
Age               110527 non-null int64
Neighbourhood     110527 non-null object
Scholarship       110527 non-null int64
Hipertension      110527 non-null int64
Diabetes          110527 non-null int64
Alcoholism        110527 non-null int64
Handcap           110527 non-null int64
SMS_received      110527 non-null int64
No-show           110527 non-null object
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB


In [4]:
med.head(10)

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   
5  9.598513e+13        5626772      F  2016-04-27T08:36:51Z   
6  7.336882e+14        5630279      F  2016-04-27T15:05:12Z   
7  3.449833e+12        5630575      F  2016-04-27T15:39:58Z   
8  5.639473e+13        5638447      F  2016-04-29T08:02:16Z   
9  7.812456e+13        5629123      F  2016-04-27T12:48:25Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   
5  2016-04-29T00:00:00Z   76          REPÚBLICA            0             1   
6  2016-04-29T00:00:00Z   23         GOIABEIRAS            0             0   
7  2016-04-29T00:00:00Z   39         GOIABEIRAS            0             0   
8  2016-04-29T00:00:00Z   21         ANDORINHAS            0             0   
9  2016-04-29T00:00:00Z   19          CONQUISTA            0             0   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0        0             0      No  
2         0           0        0             0      No  
3         0           0        0             0      No  
4         1           0        0             0      No  
5         0           0        0             0      No  
6         0           0        0             0     Yes  
7         0           0        0             0     Yes  
8         0           0        0             0      No  
9         0           0        0             0      No

**Explanation of variables that remain unclear:**

**Scholarship:** The Brazilian government gives scholarships as a subsidy for very low-income families if they have children going to school and other more specific requirements. The scholarship column refers to wheter the patient receives financial subsidies from the government (Yes=1) or not (No=0).

**Hipertension:** "Bluthochdruck" in German

**Handcap:** The handcap column refers to the number of disabilites a patient has. For example, if the patient is blind and can't walk the total is 2. 0 is used if the patient does not have a handicap. As it is misspelled, I rename it to "Handicap" below:

In [5]:
med = med.rename(columns={"Handcap": "Handicap"})

To make the column name spelling consistent, I rename the other columns too:

In [6]:
med.columns = ['Patient_ID', 'Appointment_ID', 'Gender', 'Scheduled_Day', 
              'Appointment_Day', 'Age', 'Neighbourhood', 'Scholarship', 'Hypertension',
              'Diabetes', 'Alcoholism', 'Handicap', 'SMS_Received', 'No_Show']
med.columns

Index(['Patient_ID', 'Appointment_ID', 'Gender', 'Scheduled_Day',
       'Appointment_Day', 'Age', 'Neighbourhood', 'Scholarship',
       'Hypertension', 'Diabetes', 'Alcoholism', 'Handicap', 'SMS_Received',
       'No_Show'],
      dtype='object')

In [7]:
med.dtypes

Patient_ID         float64
Appointment_ID       int64
Gender              object
Scheduled_Day       object
Appointment_Day     object
Age                  int64
Neighbourhood       object
Scholarship          int64
Hypertension         int64
Diabetes             int64
Alcoholism           int64
Handicap             int64
SMS_Received         int64
No_Show             object
dtype: object

In [8]:
med.shape

(110527, 14)

In [9]:
med.isnull().sum()

Patient_ID         0
Appointment_ID     0
Gender             0
Scheduled_Day      0
Appointment_Day    0
Age                0
Neighbourhood      0
Scholarship        0
Hypertension       0
Diabetes           0
Alcoholism         0
Handicap           0
SMS_Received       0
No_Show            0
dtype: int64

### 1.4 Inspect each column

In [10]:
# to count Patient ID's

len(med['Patient_ID'].unique().tolist())

62299

In [11]:
# to count Appointment ID's

len(med['Appointment_ID'].unique().tolist())

110527

In [12]:
# to check unique values in Gender

med.Gender.unique()

array(['F', 'M'], dtype=object)

In [13]:
# to check the time range for scheduled days

med.sort_values(by='Scheduled_Day')["Scheduled_Day"].unique()

array(['2015-11-10T07:13:56Z', '2015-12-03T08:17:28Z',
       '2015-12-07T10:40:59Z', ..., '2016-06-08T19:33:23Z',
       '2016-06-08T19:58:52Z', '2016-06-08T20:07:23Z'], dtype=object)

In [14]:
# to check the time range for appointment days and to test if information on hour of day is given

med.sort_values(by='Appointment_Day')["Appointment_Day"].unique()

array(['2016-04-29T00:00:00Z', '2016-05-02T00:00:00Z',
       '2016-05-03T00:00:00Z', '2016-05-04T00:00:00Z',
       '2016-05-05T00:00:00Z', '2016-05-06T00:00:00Z',
       '2016-05-09T00:00:00Z', '2016-05-10T00:00:00Z',
       '2016-05-11T00:00:00Z', '2016-05-12T00:00:00Z',
       '2016-05-13T00:00:00Z', '2016-05-14T00:00:00Z',
       '2016-05-16T00:00:00Z', '2016-05-17T00:00:00Z',
       '2016-05-18T00:00:00Z', '2016-05-19T00:00:00Z',
       '2016-05-20T00:00:00Z', '2016-05-24T00:00:00Z',
       '2016-05-25T00:00:00Z', '2016-05-30T00:00:00Z',
       '2016-05-31T00:00:00Z', '2016-06-01T00:00:00Z',
       '2016-06-02T00:00:00Z', '2016-06-03T00:00:00Z',
       '2016-06-06T00:00:00Z', '2016-06-07T00:00:00Z',
       '2016-06-08T00:00:00Z'], dtype=object)

In [15]:
# to check ages

med.sort_values(by='Age')["Age"].unique()

array([ -1,   0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,
        12,  13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,
        25,  26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,
        38,  39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,
        51,  52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,
        64,  65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,
        77,  78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,
        90,  91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 102, 115],
      dtype=int64)

In [16]:
med.loc[med['Age'] == -1]

Patient_ID  Appointment_ID Gender         Scheduled_Day  \
99832  4.659432e+14         5775010      F  2016-06-06T08:58:13Z   

            Appointment_Day  Age Neighbourhood  Scholarship  Hypertension  \
99832  2016-06-06T00:00:00Z   -1         ROMÃO            0             0   

       Diabetes  Alcoholism  Handicap  SMS_Received No_Show  
99832         0           0         0             0      No

In [17]:
med.loc[(med['Age'] > 100)]

Patient_ID  Appointment_ID Gender         Scheduled_Day  \
58014  9.762948e+14         5651757      F  2016-05-03T09:14:53Z   
63912  3.196321e+13         5700278      F  2016-05-16T09:17:44Z   
63915  3.196321e+13         5700279      F  2016-05-16T09:17:44Z   
68127  3.196321e+13         5562812      F  2016-04-08T14:29:17Z   
76284  3.196321e+13         5744037      F  2016-05-30T09:44:51Z   
90372  2.342836e+11         5751563      F  2016-05-31T10:19:49Z   
97666  7.482346e+14         5717451      F  2016-05-19T07:57:56Z   

            Appointment_Day  Age Neighbourhood  Scholarship  Hypertension  \
58014  2016-05-03T00:00:00Z  102     CONQUISTA            0             0   
63912  2016-05-19T00:00:00Z  115    ANDORINHAS            0             0   
63915  2016-05-19T00:00:00Z  115    ANDORINHAS            0             0   
68127  2016-05-16T00:00:00Z  115    ANDORINHAS            0             0   
76284  2016-05-30T00:00:00Z  115    ANDORINHAS            0             0   
90372  2016-06-02T00:00:00Z  102   MARIA ORTIZ            0             0   
97666  2016-06-03T00:00:00Z  115      SÃO JOSÉ            0             1   

       Diabetes  Alcoholism  Handicap  SMS_Received No_Show  
58014         0           0         0             0      No  
63912         0           0         1             0     Yes  
63915         0           0         1             0     Yes  
68127         0           0         1             0     Yes  
76284         0           0         1             0      No  
90372         0           0         0             0      No  
97666         0           0         0             1      No

In [18]:
# to show unique neighbourhoods:

med.Neighbourhood.unique()

array(['JARDIM DA PENHA', 'MATA DA PRAIA', 'PONTAL DE CAMBURI',
       'REPÚBLICA', 'GOIABEIRAS', 'ANDORINHAS', 'CONQUISTA',
       'NOVA PALESTINA', 'DA PENHA', 'TABUAZEIRO', 'BENTO FERREIRA',
       'SÃO PEDRO', 'SANTA MARTHA', 'SÃO CRISTÓVÃO', 'MARUÍPE',
       'GRANDE VITÓRIA', 'SÃO BENEDITO', 'ILHA DAS CAIEIRAS',
       'SANTO ANDRÉ', 'SOLON BORGES', 'BONFIM', 'JARDIM CAMBURI',
       'MARIA ORTIZ', 'JABOUR', 'ANTÔNIO HONÓRIO', 'RESISTÊNCIA',
       'ILHA DE SANTA MARIA', 'JUCUTUQUARA', 'MONTE BELO',
       'MÁRIO CYPRESTE', 'SANTO ANTÔNIO', 'BELA VISTA', 'PRAIA DO SUÁ',
       'SANTA HELENA', 'ITARARÉ', 'INHANGUETÁ', 'UNIVERSITÁRIO',
       'SÃO JOSÉ', 'REDENÇÃO', 'SANTA CLARA', 'CENTRO', 'PARQUE MOSCOSO',
       'DO MOSCOSO', 'SANTOS DUMONT', 'CARATOÍRA', 'ARIOVALDO FAVALESSA',
       'ILHA DO FRADE', 'GURIGICA', 'JOANA D´ARC', 'CONSOLAÇÃO',
       'PRAIA DO CANTO', 'BOA VISTA', 'MORADA DE CAMBURI', 'SANTA LUÍZA',
       'SANTA LÚCIA', 'BARRO VERMELHO', 'ESTRELINHA', 'FORTE SÃO 

In [19]:
# to count unique Neighbourhoods:

med.Neighbourhood.unique().size

81

In [20]:
# to change capital letters to title in Neighbourhood:

med["Neighbourhood"]= med["Neighbourhood"].str.title() 

In [21]:
# to check values for scholarship:

med.Scholarship.unique()

array([0, 1], dtype=int64)

In [22]:
# to check values for hypertension:

med.Hypertension.unique() 

array([1, 0], dtype=int64)

In [23]:
# to check values for diabetes:

med.Diabetes.unique() 

array([0, 1], dtype=int64)

In [24]:
# to check values for alcoholism:

med.Alcoholism.unique() 

array([0, 1], dtype=int64)

In [25]:
# to check values for handicap:

med.Handicap.unique()

array([0, 1, 2, 3, 4], dtype=int64)

In [26]:
# to count occurences of unique values for handicap:

med.groupby('Handicap').count()["Patient_ID"]

Handicap
0    108286
1      2042
2       183
3        13
4         3
Name: Patient_ID, dtype: int64

In [27]:
# to check values for SMS received:

med.SMS_Received.unique() 

array([0, 1], dtype=int64)

In [28]:
# to check values for No-Show:

med.No_Show.unique() 

array(['No', 'Yes'], dtype=object)

In [29]:
# absolute numbers No-Show:

print(med.groupby(['No_Show']).size())

No_Show
No     88208
Yes    22319
dtype: int64


In [30]:
# percentages No-Show:

med['No_Show'].value_counts(normalize=True) * 100

No     79.806744
Yes    20.193256
Name: No_Show, dtype: float64

# First Findings and Action Items:

1. The dataset contains:
       - 110527 entries in total
       - 1 dependent variable ("No-Shows") this analyses focuses on
       - 13 independent variables
       - data collected on 62299 Patients according to unique Patient ID's
       - information on 110527 appointments according to unique Appointment_ID's
       - 81 neighbourhoods
       - no missing values
<br>
2. The scheduled day dates range from the 10th November 2015 to the 8th June 2016. So data for scheduled days comprises **seven months**, as appointments are usually scheduled far in advance.<br>
<br>
3. The appointment day dates range from the 29th April 2016 to the 8th June 2016. So data for appointment days comprises **a bit more than one month**. Moreover, no information on the hour of day is provided, so this information can be dropped.<br>
<br>
4. There is one entry where **patient age equals -1**. This needs to be corrected. Furthermore, there are seven entries where **patients' age exceeds 100.**<br>
<br>
5. As the target variable is No-Show, it was necessary to find out how many patients show and do not show up to their appointments. Here the **percentage for patients not showing up is: 20.19%**. For patients showing up the percentage is: 79.81%.<br>
<br>
6. The variable types seem to be suitable except for:
       - PatientId (float)-> should be converted to int
       - Gender (object)-> should be converted to bool for machine learning purposes later on
       - ScheduledDay (object) -> should be converted to datetime object
       - AppointmentDay (object) -> should be converted to datetime object
       - No-show (object) -> should be converted to bool for machine learning purposes later on
<br>
7. It seems to be of interest to find out how long patients waited for their appointments and if their waiting time is related to wheter they show up to their appointments or not. I will create a column to display the time between scheduled day and appointment day.<br>
<br>
8. Furthermore, I am curious to figure out whether the day of the week is related to patients' show or no-show to appointments. I will create a column to display the day of the week for the appointment day.<br>

# 2. Data Cleaning

### 2.1 Change column types

In [31]:
# change type patient_id from float to int:

med['Patient_ID'] = med['Patient_ID'].astype('int64')

In [32]:
len(med['Patient_ID'].unique().tolist())

62299

In [33]:
# change type Scheduled_Day from object to datetime:

med['Scheduled_Day'] = pd.to_datetime(med['Scheduled_Day']).dt.date.astype('datetime64')

In [34]:
# change type Appointment_Day from object to datetime:

med['Appointment_Day'] = pd.to_datetime(med['Appointment_Day']).dt.date.astype('datetime64')

In [35]:
med.dtypes

Patient_ID                  int64
Appointment_ID              int64
Gender                     object
Scheduled_Day      datetime64[ns]
Appointment_Day    datetime64[ns]
Age                         int64
Neighbourhood              object
Scholarship                 int64
Hypertension                int64
Diabetes                    int64
Alcoholism                  int64
Handicap                    int64
SMS_Received                int64
No_Show                    object
dtype: object

### 2.2 Delete row with entry where age equals -1
    I assume it is a measurement error.

In [36]:
med.drop(med.loc[med['Age']<0].index, inplace=True)

### 2.3 Create new column "Day_Of_Week"
    To display the day of the week of the appointment day.

In [37]:
med["Day_Of_Week"] = med.Appointment_Day.dt.weekday_name

In [38]:
# to display which day of the week has the most appointments

days = med.groupby('Day_Of_Week').count().sort_values(by=['Day_Of_Week'],ascending=False)["Patient_ID"].to_frame()
days.rename(columns={"Patient_ID": "Day_Count"},inplace=True)
days.sort_values(by=["Day_Count"],ascending=False)

Day_Count
Day_Of_Week           
Wednesday        25867
Tuesday          25640
Monday           22714
Friday           19019
Thursday         17247
Saturday            39

### 2.4 Create new column "Waiting_Days"
    To calculate the time between the day the appointment is scheduled and the day the appointment takes place.

In [39]:
med["Waiting_Days"] = (med.Appointment_Day - med.Scheduled_Day).dt.days

In [40]:
med.sort_values(by="Waiting_Days")["Waiting_Days"].unique()

array([ -6,  -1,   0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,
        11,  12,  13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,
        24,  25,  26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,
        37,  38,  39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,
        50,  51,  52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,
        63,  64,  65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,
        76,  77,  78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,
        89,  90,  91,  92,  93,  94,  95,  96,  97,  98, 101, 102, 103,
       104, 105, 107, 108, 109, 110, 111, 112, 115, 117, 119, 122, 123,
       125, 126, 127, 132, 133, 139, 142, 146, 151, 155, 162, 169, 176,
       179], dtype=int64)

In [41]:
med.loc[(med['Waiting_Days'] < 0)]

Patient_ID  Appointment_ID Gender Scheduled_Day Appointment_Day  \
27033    7839272661752         5679978      M    2016-05-10      2016-05-09   
55226    7896293967868         5715660      F    2016-05-18      2016-05-17   
64175   24252258389979         5664962      F    2016-05-05      2016-05-04   
71533  998231581612122         5686628      F    2016-05-11      2016-05-05   
72362    3787481966821         5655637      M    2016-05-04      2016-05-03   

       Age  Neighbourhood  Scholarship  Hypertension  Diabetes  Alcoholism  \
27033   38    Resistência            0             0         0           0   
55226   19  Santo Antônio            0             0         0           0   
64175   22     Consolação            0             0         0           0   
71533   81  Santo Antônio            0             0         0           0   
72362    7     Tabuazeiro            0             0         0           0   

       Handicap  SMS_Received No_Show Day_Of_Week  Waiting_Days  
27033         1             0     Yes      Monday            -1  
55226         1             0     Yes     Tuesday            -1  
64175         0             0     Yes   Wednesday            -1  
71533         0             0     Yes    Thursday            -6  
72362         0             0     Yes     Tuesday            -1

    I assume that the days patients waited for their appointments cannot be negative.
    Therefore, I remove all rows with entries less than 0.

In [42]:
med.drop(med.loc[med['Waiting_Days']<0].index, inplace=True)

In [43]:
med.shape

(110521, 16)

In [44]:
med.head()

Patient_ID  Appointment_ID Gender Scheduled_Day Appointment_Day  Age  \
0   29872499824296         5642903      F    2016-04-29      2016-04-29   62   
1  558997776694438         5642503      M    2016-04-29      2016-04-29   56   
2    4262962299951         5642549      F    2016-04-29      2016-04-29   62   
3     867951213174         5642828      F    2016-04-29      2016-04-29    8   
4    8841186448183         5642494      F    2016-04-29      2016-04-29   56   

       Neighbourhood  Scholarship  Hypertension  Diabetes  Alcoholism  \
0    Jardim Da Penha            0             1         0           0   
1    Jardim Da Penha            0             0         0           0   
2      Mata Da Praia            0             0         0           0   
3  Pontal De Camburi            0             0         0           0   
4    Jardim Da Penha            0             1         1           0   

   Handicap  SMS_Received No_Show Day_Of_Week  Waiting_Days  
0         0             0      No      Friday             0  
1         0             0      No      Friday             0  
2         0             0      No      Friday             0  
3         0             0      No      Friday             0  
4         0             0      No      Friday             0

# Observations after cleaning:

**Age:** The average age of patients is 37 while 75% of patients are aged below 55. There is also patients aged higher than 100, which can be derived from the maximum age value (115).<br>
<br>
**Waiting Days:** The average of days patients wait for their appointment is 15. However, the maximum value for patients waiting for an appointment is 115. 25% of all patients show 0 waiting days so the appointment was scheduled and performed on the same day.<br>
<br> 
**SMS Received:** 75% of all patients received an SMS prior to their appointment.<br>
<br> 
**Day Of Week:** The data seems to be almost evenly spread among workdays with Friday and Thursday showing slightly less appointments. There are 39 appointments on Saturday:<br> 
<br> 
Wednesday: 25867<br> 
Tuesday: 25640<br> 
Monday: 22715<br> 
Friday: 19019<br> 
Thursday: 17247<br> 
Saturday: 39<br> 

In [45]:
med.to_pickle('med.pkl')